scripts/quests/otherAreas/Fishermans_Heart.lua

```diff
@@ -0,0 +1,147 @@
+        },
+    },
+
+    -- Quest available, skill >= 20, Windurst Fame >= 2
+    {
+        check = function(player, status, vars)
+            return
+                status == xi.questStatus.QUEST_AVAILABLE and
+                player:getFameLevel(xi.fameArea.WINDURST) >= 2 and
+                player:getCharSkillLevel(xi.skill.FISHING) >= 200
+        end,
+
+        [xi.zone.MHAURA] =
+        {
+            ['Katsunaga'] =
+            {
+                onTrigger = function(player, npc)
+                    return quest:progressEvent(192, xi.item.GUGRU_TUNA)
+                end,
+            },
+
+            onEventFinish =
+            {
+                [192] = function(player, csid, option, npc)
+                    quest:begin(player)
+                end,
+            },
+        },
+    },
+
+    -- Quest Accepted
+    {
+        check = function(player, status, vars)
+            return status == xi.questStatus.QUEST_ACCEPTED
+        end,
+
+        [xi.zone.MHAURA] =
+        {
+            ['Katsunaga'] =
+            {
+                onTrigger = function(player, npc)
+                    return quest:progressEvent(194, xi.item.GUGRU_TUNA)
+                end,
+
+                onTrade = function(player, npc, trade)
+                    if npcUtil.tradeHasExactly(trade, xi.item.GUGRU_TUNA) then
+                        local fishingStats = player:getFishingStats() or {}
+                        return quest:progressEvent(193, {
+                            [1] = fishingStats['fishLinesCast'] or 0,
+                            [2] = fishingStats['fishReeled'] or 0,
+                            [3] = fishingStats['fishLongestLength'] or 0,
+                            [4] = fishingStats['fishLongestId'] or 0,
+                            [5] = fishingStats['fishHeaviestWeight'] or 0,
+                            [6] = fishingStats['fishHeaviestId'] or 0,
+                        })
+                    end
+                end,
+            },
+
+            onEventFinish =
+            {
+                [193] = function(player, csid, option, npc)
+                    player:confirmTrade()
+                    quest:complete(player)
+                end,
+            },
+        },
+    },
+
+    -- Quest Complete
+    {
+        check = function(player, status, vars)
+            return status == xi.questStatus.QUEST_COMPLETED
+        end,
+
+        [xi.zone.MHAURA] =
+        {
+            ['Katsunaga'] =
+            {
+                onTrigger = function(player, npc)
+                    return quest:progressEvent(197, player:getFishingCatchCount() or 0) -- Number of fish types caught
+                end,
+            },
+
+            onEventUpdate =
+            {
+                [197] = function(player, csid, option, npc)
+                    if option == 33554432 then
+                        -- 'Catching Fish'
+                        local fishingStats = player:getFishingStats() or {}
```

Comment:
Haven't gotten there yet, but I'd rather break out if fishing stats are non-existent.  updateEvent can take 0 params, so no need to pass an empty table.  Also, is this data returned always the same, and could we consolidate it to always return what we need in the format needed?

Reply:
Good to go on the first part, but I'm not sure I understand the second question?  the code for `getFishingStats()` checks the contents of PChar->m_fishHistory, which is a fixed struct.  If the C++ function fails, it returns a table of 0s instead with the same labels.  

```lua
 local fishingStats = player:getFishingStats()
 if fishingStats ~= nil then
     player:updateEvent({
         [0] = player:getFishingCatchCount() or 0,
         [1] = fishingStats['fishLinesCast'] or 0,
         [2] = fishingStats['fishReeled'] or 0,
         [3] = fishingStats['fishLongestLength'] or 0,
         [4] = fishingStats['fishLongestId'] or 0,
         [5] = fishingStats['fishHeaviestWeight'] or 0,
         [6] = fishingStats['fishHeaviestId'] or 0,
     })
 else
     player:updateEvent(0, 0, 0, 0, 0, 0, 0)
 end